fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/services/MembershipIndexManager.java

```diff
@@ -0,0 +1,630 @@
+            " FROM membership m" +
+            " WHERE m.source_id = :sourceId" +
+                " AND m.end_time = :noEndTime" +
+                " AND m.subject_id = :subjectId" +
+                " AND m.property = :property" +
+                " AND m.object_id = :objectId";
+
+    private static final String END_MEMBERSHIP_IN_TX =
+            "UPDATE membership_tx_operations" +
+            " SET end_time = :endTime" +
+            " WHERE source_id = :sourceId" +
+                " AND end_time = :noEndTime" +
+                " AND subject_id = :subjectId" +
+                " AND property = :property" +
+                " AND object_id = :objectId";
+
+    private static final String CLEAR_ENTRY_IN_TX =
+            "DELETE FROM membership_tx_operations" +
+            " WHERE source_id = :sourceId" +
+                " AND tx_id = :txId" +
+                " AND subject_id = :subjectId" +
+                " AND property = :property" +
+                " AND object_id = :objectId" +
+                " AND operation = :operation" +
+                " AND force_flag IS NULL";
+
+    private static final String CLEAR_ALL_ADDED_FOR_SOURCE_IN_TX =
+            "DELETE FROM membership_tx_operations" +
+            " WHERE source_id = :sourceId" +
+                " AND tx_id = :txId" +
+                " AND operation = :addOp";
+
+    // Add "delete" entries for all existing membership from the given source, if not already deleted
+    private static final String END_EXISTING_FOR_SOURCE =
+            "INSERT INTO membership_tx_operations" +
+            " (subject_id, property, object_id, source_id, start_time, end_time, tx_id, operation)" +
+            " SELECT subject_id, property, object_id, source_id, start_time, :endTime, :txId, :deleteOp" +
+            " FROM membership m" +
+            " WHERE source_id = :sourceId" +
+                " AND end_time = :noEndTime" +
+                " AND NOT EXISTS (" +
+                    " SELECT TRUE" +
+                    " FROM membership_tx_operations mtx" +
+                    " WHERE mtx.subject_id = m.subject_id" +
+                        " AND mtx.property = m.property" +
+                        " AND mtx.object_id = m.object_id" +
+                        " AND mtx.source_id = m.source_id" +
+                        " AND mtx.operation = :deleteOp" +
+                    ")";
+
+    private static final String DELETE_EXISTING_FOR_SOURCE_AFTER =
+            "INSERT INTO membership_tx_operations" +
+            " (subject_id, property, object_id, source_id, start_time, end_time, tx_id, operation, force_flag)" +
+            " SELECT subject_id, property, object_id, source_id, start_time, end_time, :txId, :deleteOp, :forceFlag" +
+            " FROM membership m" +
+            " WHERE m.source_id = :sourceId" +
+                " AND (m.start_time >= :startTime" +
+                " OR m.end_time >= :startTime)";
```

Comment:
Yes, that's right. It is to deal with obnoxious history rewriting from modifying mutable head versions of DirectContainers. Sorry to do this to you, but consider the following timeline, in a repo with autoversioning turned off:

1) Create membership resc `coll1` and direct container `dc1` with membership relation `ex:hasMember1`, at time T0.
2) At time T1, resource `dc1/mem1` is created. This adds an entry to the index:
`<coll1> ex:hasMember1 <dc1/mem1>` which spans T1 to forever.
3) At time T2, `dc1` is changed to have relation `ex:hasMember2`. The index is now:
`<coll1> ex:hasMember2 <dc1/mem1>` which spans T1 to forever (the old entry is deleted since its end time was >= T2)
4) At T3, create V1 of `dc1`. No changes to the index.
5) At time T4, `dc1` is changed to have relation `ex:hasMember3`. The index is now:
`<coll1> ex:hasMember2 <dc1/mem1>` which spans T1 to T4 (T3/V1 locked ex:hasMember2 in, which changes at T4)
`<coll1> ex:hasMember3 <dc1/mem1>` which spans T4 to forever.
6) At T5, `dc1` is changed back to relation `ex:hasMember1`. History in the index is editted:
`<coll1> ex:hasMember2 <dc1/mem1>` which spans T1 to T5 (note, end time changed)
`<coll1> ex:hasMember1 <dc1/mem1>` which spans T5 to forever.

The query in question allows the last few states to be rewritten. Alternatively, I could regenerate the full history of membership produced by `dc1` whenever its modified, which I had earlier in the PR. Its simpler, but obviously costly in scenarios where clients create versions of DCs and modify them.

Reply:
Thanks for the explanation and deep consideration of scenarios.